numpy/core/src/multiarray/einsum_sumprod.c.src

```diff
@@ -250,244 +236,198 @@ static void
-            b = _mm_add_ps(a, _mm_load_ps(data_out+@i@));
-            _mm_store_ps(data_out+@i@, b);
-/**end repeat2**/
-            data1 += 8;
-            data_out += 8;
-        }
-
-        /* Finish off the loop */
-        if (count > 0) {
-            goto finish_after_unrolled_loop;
+    #ifndef NPY_HAVE_NEON
+    const int is_aligned = EINSUM_IS_ALIGNED(data1) && EINSUM_IS_ALIGNED(data_out);
+    #else
+    // ARM/Neon don't have instructions for aligned memory access
+    const int is_aligned = 0;
+    #endif
+    const int vstep = npyv_nlanes_@sfx@;
+    const npyv_@sfx@ va_scalar = npyv_setall_@sfx@(a_scalar);
+
+    /**begin repeat2
+     * #cond = if(is_aligned), else#
+     * #ld = loada, load#
+     * #st = storea, store#
+     */
+    @cond@ {
+    #if @unroll_by@ == 4
+        const npy_intp vstepx4 = vstep * 4;
+        for (; count >= vstepx4; count -= vstepx4, data1 += vstepx4, data_out += vstepx4) {
+            /**begin repeat3
+             * #i = 0, 1, 2, 3#
+             */
+            npyv_@sfx@ b@i@ = npyv_@ld@_@sfx@(data1 + vstep * @i@);
+            npyv_@sfx@ c@i@ = npyv_@ld@_@sfx@(data_out + vstep * @i@);
+            /**end repeat3**/
+            /**begin repeat3
+             * #i = 0, 1, 2, 3#
+             */
+            npyv_@sfx@ abc@i@ = npyv_muladd_@sfx@(va_scalar, b@i@, c@i@);
+            /**end repeat3**/
+            /**begin repeat3
+             * #i = 0, 1, 2, 3#
+             */
+            npyv_@st@_@sfx@(data_out + vstep * @i@, abc@i@);
+            /**end repeat3**/
         }
-        else {
-            return;
+    #elif @unroll_by@ == 2
+        const npy_intp vstepx2 = vstep * 2;
+        for (; count >= vstepx2; count -= vstepx2, data1 += vstepx2, data_out += vstepx2) {
+            npyv_@sfx@ b0 = npyv_@ld@_@sfx@(data1);
+            npyv_@sfx@ b1 = npyv_@ld@_@sfx@(data1 + vstep);
+            npyv_@sfx@ c0 = npyv_@ld@_@sfx@(data_out);
+            npyv_@sfx@ c1 = npyv_@ld@_@sfx@(data_out + vstep);
+            npyv_@sfx@ abc0 = npyv_muladd_@sfx@(va_scalar, b0, c0);
+            npyv_@sfx@ abc1 = npyv_muladd_@sfx@(va_scalar, b1, c1);
+            npyv_@st@_@sfx@(data_out, abc0);
+            npyv_@st@_@sfx@(data_out + vstep, abc1);
         }
+    #endif
     }
-#elif EINSUM_USE_SSE2 && @float64@
-    value0_sse = _mm_set1_pd(value0);
-
-    /* Use aligned instructions if possible */
-    if (EINSUM_IS_SSE_ALIGNED(data1) && EINSUM_IS_SSE_ALIGNED(data_out)) {
-        /* Unroll the loop by 8 */
-        while (count >= 8) {
-            count -= 8;
-
-/**begin repeat2
- * #i = 0, 2, 4, 6#
+    /**end repeat2**/
+    npyv_cleanup();
+#endif // NPYV check for @type@
+/**
+ * Unroll by four scalars in case of:
+ *  - The SIMD width is higher than 128bit since we unroll by x2/x4
+ *    and that may lead to performance loss on small arrays.
+ *  - To give the change to the compiler to
+ *    auto-vectorize in case of NPYV wasn't available.
```

Comment:
I don't quite understand this comment and the `!defined(NPY_DISABLE_OPTIMIZATION) && (!@NPYV_CHK@ || NPY_SIMD > 128)` condition below.

Given this comment and condition appears in 8 places, can you hoist it to the top of the file as something like
```
/* similar comment to before */
#define EINSUM_UNROLL_4_SCALARS !defined(NPY_DISABLE_OPTIMIZATION) && (!@NPYV_CHK@ || NPY_SIMD > 128)
```

Reply:
That comment should explaind the statment :
`!defined(NPY_DISABLE_OPTIMIZATION) && (!@NPYV_CHK@ || NPY_SIMD > 128)`

Anway, we can clearfy the above tokens as following:
1. `NPY_DISABLE_OPTIMIZATION` is a compiler definition initlized by build option `--disable-optimization`.
2. `@NPYV_CHK@` substituted by template var during template iteration to the following values:
   - for **float**: `NPY_SIMD`
   - for **double**: `NPY_SIMD_F64`
   - for all other types : `0`
3. `NPY_SIMD` hold the SIMD witdth in bits or `0` if `NPYV`:
   - doesn't support the current arch
   - the supported SIMD extentions aren't enabled by infrastrcture.
4. `NPY_SIMD_F64`:  `1` if double-prescion operations is supported, otherwise `0`.
   Currently it set `0` only for `NEON` on `Armv7/A32`.

Now, in the light of above, the un-roll(4xScalars) branch should be enabled if:
 1.  `!defined(NPY_DISABLE_OPTIMIZATION)`:  optimization isn't disabled.
 2. (1.) and `!`@NPYV_CHK@` : `NPYV` is not enabled for the itereted type. (`0`)
 3. (1.) and `NPY_SIMD` > 128: if `NPYV` is enabled with SIMD width larger than `128` e.g. `256`(AVX2) or `512`(AVX512)
    Since the main SIMD loop un-rolled by `NPY_SIMD_WIDTH`*(2|4),
    which may lead to un-vectorize the remained scalars in bytes range  <=`NPY_SIMD_WIDTH`*(2|4).
    I think we can block the unroll branch for this case since modern compilers will auto-vectorized anyway.

Reply:
it also equivalent to:
```C
(!defined(NPY_DISABLE_OPTIMIZATION) && !@NPYV_CHK@) || NPY_SIMD > 128
```
since `NPY_DISABLE_OPTIMIZATION` affect on availability of `NPY_SIMD`.

Reply:
defined a common macro.